Expand search box horizontally when text grows long

sinistersnare opened the pull request:
Hello there! This is my first contribution. Let me know if I should add a test for this (somehow?) or do anything else required. My goal is to learn the codebase a bit and the development workflow here.

<!--
Thanks for contributing to JupyterLab!
Please fill out the following items to submit a pull request.
See the contributing guidelines for more information:
https://github.com/jupyterlab/jupyterlab/blob/main/CONTRIBUTING.md
-->

## References

Should close #15049. Thanks to #15080 for the work to jump off of!

<!-- Note issue numbers this pull request addresses (should be at least one, see contributing guidelines above). -->

<!-- Note any other pull requests that address this issue and how this pull request is different. -->

## Code changes

<!-- Describe the code changes and how they address the issue. -->

1. I fixed a typo in docs
2. I added a common hack to naturally grow `<textarea>` elements based on their contents. This utilizes some CSS hackery, but it is [well documented](https://css-tricks.com/auto-growing-inputs-textareas/)

## User-facing changes

Before: [jupyter_search_width_before.webm](https://github.com/jupyterlab/jupyterlab/assets/2153833/ad84acf2-d047-485b-9af6-a14158872a84)

After: [jupyter_search_width_after.webm](https://github.com/jupyterlab/jupyterlab/assets/2153833/558ad790-f7eb-4ed9-bc38-8a74e0ab0a5f)

(the viewer isnt showing up for me and idk how to make it do that)


<!-- Describe any visual or user interaction changes and how they address the issue. -->

<!-- For visual changes, include before and after screenshots or GIF/mp4/other video demo here. -->

## Backwards-incompatible changes

No API changes here.

<!-- Describe any backwards-incompatible changes to JupyterLab public APIs. -->

sinistersnare commented on packages/documentsearch/src/searchview.tsx:
```diff
@@ -77,13 +78,20 @@ export interface ISearchKeyBindings {
+        if (element.parentNode && element.parentNode instanceof HTMLElement) {
+          element.parentNode.dataset.value = longest;
+        }
+        // element.parentNode?.dataset.value = longest;
```
dead code here, will delete along with any other possible changes that need to be made.

krassowski reviewed (commented):
Thank you for starting this, it looks like a great solution. It should probably have a brief comment documenting how it works in the CSS code because it relies on a bit of a less-known technique.

> should add a test for this (somehow?)

Yes, that would be appreciated. It could be build upon:

https://github.com/jupyterlab/jupyterlab/blob/0f178d3094f07f4474dc171a9493effbf0b38e90/galata/test/jupyterlab/search.test.ts#L29-L39

Some changes will be needed to limit the growth at certain point, or to ensure that other UI elements do not get squished. In particular currently a very long text will make some buttons unreadable/unclickable:

![image](https://github.com/jupyterlab/jupyterlab/assets/5832902/be1203c7-5ea5-458c-9c24-bcedbc9337e8)

![image](https://github.com/jupyterlab/jupyterlab/assets/5832902/1ae84800-b8c6-4ca2-abe4-9bac9d8d65cf)

sinistersnare commented:
Hey there, fixed up some of the CSS to work a bit better. Now the buttons next to the search boxes do not shrink, and I added some docs.

Still working on the tests, it seems like the failures are just screenshot failures. Is the protocol to just update them?

In terms of adding a new test I can do that, but would be more comfortable doing that after I get all the tests passing.

sinistersnare commented:
please update galata snapshots

sinistersnare commented on packages/documentsearch/style/base.css:
```diff
@@ -32,6 +32,7 @@
 .jp-DocumentSearch-overlay button {
   background-color: var(--jp-toolbar-background);
   outline: 0;
+  margin-top: 1.5px;
```
I tried to address the spacing of the buttons with this. I am not so strong with CSS so if there is a better solution let me know!

fcollonval commented:
Kicking the CI following the generation of the new snapshots

krassowski commented:
please update galata snapshots

krassowski reviewed (approved):
Thank you @sinistersnare!

krassowski commented:
@meeseeksdev please backport to 4.0.x

sinistersnare commented:
Hey there, I appreciate your help getting this through! Thanks!
